### ISSUE TITLE
Inventory Specification without files impossible?

### ISSUE THREAD

**Issue by SirUli** [ansible/ansible-runner#947]
Hi,

i believe that the issue #642 was accidentally closed as the underlying issue isn't resolved. The [code](https://github.com/ansible/ansible-runner/blob/devel/ansible_runner/interface.py#L148-L154) says:

> :param str or dict or list inventory: Overrides the inventory directory/file (supplied at ``private_data_dir/inventory``) with
>        a specific host or list of hosts. This can take the form of:
>            - Path to the inventory file in the ``private_data_dir``
>            - Native python dict supporting the YAML/json inventory structure
>            - A text INI formatted string
>            - A list of inventory sources, or an empty list to disable passing inventory

What works is the first (path) and the last option (list of sources or empty) but not the native dictionary or the ini-formatted string as already declared in the previous incident. Which makes sense since ansible-runner doesn't generate a file out of these somewhere.

In the [code](https://github.com/ansible/ansible-runner/blob/devel/ansible_runner/config/runner.py#L289-L297) i then saw that there isn't really a way of bringing in a proper dictionary. What currently works is to construct the list of hosts as a simple string encapsulated in a list:

- One Host: ``inventory = ["host1,"]``
- Multiple Hosts: ``inventory = ["host1,host2"]``

since that slips through all checks for physical locations of the file and is "proper" ansible syntax to be appended to "-i". But it is very limited since i cannot bring in groups, host variables etc. I assume that wasn't the original idea of the comment above ;)

Why i need that? I use runner in the python native approach from a automation system. This approach requires - besides the directory for the generated artifacts - no further directory structures etc since everything else can be brought in via the different variables.

**Comment by Akasurde** [comment]
@SirUli You need to pass the inventory JSON as - 

```
{"all": {"hosts": {"host1": null, "host2": null}}, "db": {"hosts": {"hosts3": null}}}
```

So your script might look like this - 

```python
...
inventory_json_str = '{"all": {"hosts": {"vm1": null}}}'
r = ansible_runner.run(
    host_pattern='all',
    module='ping',
    inventory=inventory_json_str,
)
...
```

If you have hostvars and groupvars,

```python
inventory ={
  "all": {
    "hosts": {
      "vm1": None,
      "vm2": None
    }
  },
  "db": {
      "hosts": {
          "vm3": {
              "index": "index.php"
          }
      },
      "vars": {
          "webroot": "/web/public_html",
          "index": "index.html"
      }
  }
}
```

**Comment by Akasurde** [comment]
I will go ahead and close this issue since there is no input from the original reporter. Please re-open the issue if the problem persists. Thanks.

**Comment by PrieJos** [comment]
Hello,

@Akasurde unfortunately your proposed solution does not work. 

Cheers
Pri

**Comment by Akasurde** [comment]
@PrieJos Can you please open a new issue with your findings? Thanks,

**Comment by PrieJos** [comment]
Hello,

@Akasurde Thanks for the reply. Indeed I updated to latest runner version (2.3.1) and change slightly the way we are calling runner and the it worked as you suggested. Indeed all I did was to use this helper `ansible_runner.interface.run(**kwargs)` as described at [Using Runner as a Python Module Interface to Ansible](https://ansible-runner.readthedocs.io/en/stable/python_interface/#using-runner-as-a-python-module-interface-to-ansible). 

Thanks so much and really sorry for the extra hassle. My fault :disappointed: 

Cheers
Pri
